manifests/install.pp

```diff
@@ -68,33 +74,69 @@
       }
     }
   } else {
+    include '::archive'
+
+    $basefilename = "zookeeper-${ensure}.tar.gz"
+    $package_url = "${mirror_url}/zookeeper/zookeeper-${ensure}/${basefilename}"
+    $checksum_url = "http://www-us.apache.org/dist/zookeeper/zookeeper-${ensure}/${basefilename}.sha1"
+    $extract_path = "${install_dir}-${ensure}"
+
+    if ($manual_clean == undef) {
+      $clean = versioncmp($ensure, '3.4') ? {
+        '-1'    => true,
+        default => false,
+      }
+    } else {
+      $clean = $manual_clean
+    }
+
     package { ['zookeeper','zookeeperd']:
       ensure => absent
     }
 
     file { $install_dir:
       ensure => link,
-      target => "${install_dir}-${ensure}"
-    }
-
-    exec { 'download-zk-package':
-      command => "/usr/bin/wget -O /tmp/zookeeper-${ensure}.tar.gz ${package_mirror}/zookeeper-${ensure}/zookeeper-${ensure}.tar.gz",
-      creates => "/tmp/zookeeper-${ensure}.tar.gz"
+      target => $extract_path
     }
 
-    exec { 'install-zk-package':
-      command => "/bin/tar -xvzf /tmp/zookeeper-${ensure}.tar.gz -C /opt",
-      creates => "${install_dir}-${ensure}/zookeeper-${ensure}.jar",
+    file { $package_dir:
+      ensure  => directory,
+      owner   => 'zookeeper',
+      group   => 'zookeeper',
       require => [
-        Exec['download-zk-package']
-      ]
+        Group['zookeeper'],
+        User['zookeeper'],
+      ],
     }
-    
-    file { "${install_dir}-${ensure}":
+
+    file { $extract_path:
       ensure  => directory,
-      recurse => true,
       owner   => 'zookeeper',
-      group   => 'zookeeper'
+      group   => 'zookeeper',
+      require => [
+        Group['zookeeper'],
+        User['zookeeper'],
+      ],
+    }
+
+    archive { "${package_dir}/${basefilename}":
```

Comment:
Should this resource `notify` the service resource? (Or else, should the service resource subscribe to this resource). That way if the version is updated and zk is reinstalled, zk will be *optionally* restarted.